Bug 1333165 - after creating from a template in console, no idea what values generated parameters got

jhadvig opened the pull request:
Showing the generated parameters on the next-step page
Wasnt fully sure if we want to put the params at the beginning or the end, so attaching both versions + one that failed to create resources:
![screenshot-11](https://cloud.githubusercontent.com/assets/1668218/15613369/7512e9ae-2433-11e6-98ac-6365e88859f9.png)
![screenshot-10](https://cloud.githubusercontent.com/assets/1668218/15613372/77beaabc-2433-11e6-9cbb-486d874d28a5.png)
This one is a bit zoomed out...
![screenshot-9](https://cloud.githubusercontent.com/assets/1668218/15613376/7a69ff0a-2433-11e6-8bcc-988fd9c9231e.png)

@jwforres PTAL

jwforres commented:
definitely prefer having them on the bottom, not sure how i feel about how much space the table takes up, but I know in some cases the parameter could contain a very long string, or even a multiline string once we support that correctly.

I feel like it needs some sentence there above the table, something like "These parameters were generated automatically and often include things like passwords.  If you will need to reference these values later, copy them to a safe location."

cc @spadgett @rhamilto @ajacobs21e

jhadvig commented:
@jwforres what about: "These parameters often include things like passwords. If you will need to reference these values later, copy them to a safe location. Some of these parameters were generated automatically."

![screenshot-16](https://cloud.githubusercontent.com/assets/1668218/15652510/f8c782fa-2686-11e6-83df-bd01117bc531.png)

Reorganized the sentence a bit since not all the params are generated automatically but dont wanted to end up with something like : "Some of these parameters were generated automatically and often include things like passwords. If you will need to reference these values later, copy them to a safe location."  :anguished:

ajacobs21e commented:
I like your wording @jhadvig. Is there a more descriptive heading we can use? "Used parameters" makes me think they're not available (but I'm not familiar with the domain). 'app parameters'? 'applied parameters'? 

Is there a way now to find these parameters in the UI? If not, will there be one day?

jhadvig commented:
@ajacobs21e the only way how to find out these parameters, is to know which resources are referencing in the template, and after the resource is created user can find them in the `Environment` tab of the resource page.
I see your point with the 'Used parameters', but I wouldn't go with 'App paramaters' or 'Application parameters' since template is not necessarily tied to creating application, but rather all kinds of components - HPA, Jobs, ImageStream, ... which are not really an application but a component. For that reason I would prefer 'Used' or 'Applied Parameters', but they are kinda the same :-P

@jwforres @spadgett thoughts ?

jwforres commented:
Yep like your wording better @jhadvig

As far as finding them later, parameters don't even necessarily mean
environment variables.  Parameters can end up being used for all kinds of
things like which image to use, which source repository, etc.  So we can't
really direct them where to look for them later unfortunately.   Typically
the ones that are generated are being used in environment variables, or
secrets.

Hmm, i think i prefer Applied Parameters, or Processed Parameters ?
@bparees any opinion here?

On Tue, May 31, 2016 at 4:33 AM, Jakub Hadvig <notifications@github.com>
wrote:

> @ajacobs21e <https://github.com/ajacobs21e> the only way how to find out
> these parameters, is to know which resources are referencing in the
> template, and after the resource is created user can find them in the
> Environment tab of the resource page.
> I see your point with the 'Used parameters', but I wouldn't go with 'App
> paramaters' or 'Application parameters' since template is not necessarily
> tied to creating application, but rather all kinds of components - HPA,
> Jobs, ImageStream, ... which are not really an application but a component.
> For that reason I would prefer 'Used' or 'Applied Parameters', but they are
> kinda the same :-P
>
> @jwforres <https://github.com/jwforres> @spadgett
> <https://github.com/spadgett> thoughts ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/48#issuecomment-222625402>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABZk7XdzKUNVeyieOL2Bh3JozP7eVQSUks5qG_JxgaJpZM4Iooe3>
> .
>

spadgett commented:
> Hmm, i think i prefer Applied Parameters, or Processed Parameters ?

Is just "Parameters" too general?

bparees commented:
or Parameter Values ?

jhadvig commented:
Applied Parameter Values ?
@jwforres @spadgett @bparees

ajacobs21e commented:
I like applied parameter values

jhadvig commented:
@jwforres PTAL

rhamilto commented:
The table seems a bit jarring on this page, but it seems like the correct UI presentation since it is how we display name/value pairs elsewhere in the console.  Maybe dropping the white background color on the table cells would help (see attached)?

![no-white-cells](https://cloud.githubusercontent.com/assets/895728/15754568/abe5eafc-28c6-11e6-8da9-a01e455c2978.png)

Should the passwords be obscured by a click to show link (e.g., Show password)?  It seems like simply outputting passwords to the page is inconsistent with what we do in other password instances (although I see we output MYSQL_PASSWORD values on the environment tab of deployments (see attached)).  Hmm.

![screen shot 2016-06-02 at 1 23 04 pm](https://cloud.githubusercontent.com/assets/895728/15754220/25d55584-28c5-11e6-8667-f56c9ca492c1.PNG)

jwforres commented:
we dont know whats a password / secret key / etc other than checking the
name for "password" :-\

On Thu, Jun 2, 2016 at 1:34 PM, Robb Hamilton <notifications@github.com>
wrote:

> The table seems a bit jarring on this page, but it seems like the correct
> UI presentation since it is how we display name/value pairs elsewhere in
> the console. Maybe dropping the white background color on the table cells
> would help (see attached)?
>
> [image: no-white-cells]
> <https://cloud.githubusercontent.com/assets/895728/15754568/abe5eafc-28c6-11e6-8da9-a01e455c2978.png>
>
> Should the passwords be obscured by a click to show link (e.g., Show
> password)? It seems like simply outputting passwords to the page is
> inconsistent with what we do in other password instances (although I see we
> output MYSQL_PASSWORD values on the environment tab of deployments (see
> attached)). Hmm.
>
> [image: screen shot 2016-06-02 at 1 23 04 pm]
> <https://cloud.githubusercontent.com/assets/895728/15754220/25d55584-28c5-11e6-8667-f56c9ca492c1.PNG>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/48#issuecomment-223364600>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABZk7Udq3chm-nNhnwzHFov0ajgzMTB8ks5qHxQ_gaJpZM4Iooe3>
> .
>

jhadvig commented:
@rhamilto thanks, really like how just simple change of bg color does make difference :+1: so I've updated the PR with proposed change.
Unfortunately as @jwforres mentioned we can't tell what is and what isn't the password fur sure, just assume it, but I don't think we want to walk that path.
Though it would be good to be consistent on how we display passwords in console.

rhamilto commented:
LGTM.  

Understood about the password.  What if we made every param a click-to-show in this context?  That's sort of yucky, but at least it's not revealing passwords?  Could even add a "Show all" link to avoid multiple clicks?  @jwforres, thoughts?

rhamilto commented on app/styles/_core.less:
```diff
@@ -1029,6 +1029,12 @@ h1 small.meta, .build-config-summary .meta {
   color: @gray-light;
 }
 
+.no-background {
```
A "global" fix that would address this any time a .table appears on a gray background would be:

.surface-shaded .table {
  background-color: transparent;
}

Plus, there is then no need to add an additional class to the table.

jhadvig commented:
@rhamilto I like that :+1:

jwforres commented:
ok I discussed this with dan and because templates can actually create Secrets, we shouldn't make these visible by default.  Honestly the simplest solution is to have the Header and the message that explains everything and then have a "Show parameters" link that shows the entire table.  WDYT?

rhamilto commented:
>  Honestly the simplest solution is to have the Header and the message that explains everything and then have a "Show parameters" link that shows the entire table. WDYT?

I like the practicality of that solution, but share @jhadvig's concern that the parameters then will be easy to overlook.  How important is it the user sees the parameters?

jhadvig commented:
@rhamilto attaching the screens for the change that @jwforres proposed

![screenshot-39](https://cloud.githubusercontent.com/assets/1668218/15786146/643e6292-29bb-11e6-8f94-3effe6d5f53a.png)
![screenshot-40](https://cloud.githubusercontent.com/assets/1668218/15786149/66d8a0da-29bb-11e6-8e11-440a63c18c9a.png)

But would be good to somehow highlight the link. 

What about center the link ? Or change the link to a btn ? Mooore screeeens :grin:  : 
Proposal 1
![screenshot-41](https://cloud.githubusercontent.com/assets/1668218/15786537/798f4f42-29bd-11e6-8654-6d59f20fc638.png)
Proposal 2
![screenshot-44](https://cloud.githubusercontent.com/assets/1668218/15786545/81995e76-29bd-11e6-97ad-1b32af6adcc8.png)
Proposal 3
![screenshot-45](https://cloud.githubusercontent.com/assets/1668218/15786547/83815130-29bd-11e6-8b4f-012db8957d98.png)
Proposal 4
![screenshot-46](https://cloud.githubusercontent.com/assets/1668218/15786612/dd30ae38-29bd-11e6-9822-2379b05b5067.png)
Proposal 5
![screenshot-47](https://cloud.githubusercontent.com/assets/1668218/15786613/de3b26c8-29bd-11e6-9515-414bf76f1d7d.png)

Where I like the 1 and 5 most. Or do you have other proposals ?

rhamilto commented:
Minor nit:  add a period after "automatically".

Hmm.  I personally don't like centering since English reads from left to right.  

The button does draw the eye, although the blue one is a bit too strong.  So I suggest either sticking with the blue link left aligned or the gray button left aligned.  Opinions @spadgett, @benjaminapetersen, @sg00dwin ?

benjaminapetersen commented:
@rhamilto @jhadvig @jwforres I like the idea to hide them with a link to show, but I would not hide the table... just use `************` over the values, and click `show` to reveal.  

quick mock:
![show-values](https://cloud.githubusercontent.com/assets/280512/15787384/207a8106-2990-11e6-8c8a-734d29206c31.png)

sg00dwin commented:
I prefer @benjaminapetersen idea to show the table with ************** and a Show values link.

jwforres commented:
ok, sounds like thats the preference then :)  its fine with me

On Fri, Jun 3, 2016 at 2:17 PM, Steve Goodwin <notifications@github.com>
wrote:

> I prefer @benjaminapetersen <https://github.com/benjaminapetersen> idea
> to show the table with ************** and a Show values link.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/48#issuecomment-223653733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABZk7ahZ4RVAnTsUN9Rt4oOgKFFU2Na8ks5qIG-0gaJpZM4Iooe3>
> .
>

spadgett commented:
I think we should know what values were generated... How about a message like

> Parameters FOO, BAR, and BAZ have generated values. <a href="">Show values</a>

where clicking would show the table.

jhadvig commented:
@spadgett you mean to hide the table by default ?

I would go with @benjaminapetersen proposal by hiding the values, but I dont get the `hidden` field in the code sample. Shall the generated parameters be hidden by default and rest will be visible?

spadgett commented:
It seems to show the same values I just entered starred out. The table doesn't tell me anything new unless I click the show values link, and it might be confusing if none of the values are actually sensitive.

I feel like a message saying these variables were generated with a link to see generated values is more informative.

jwforres commented:
@liggitt @bparees @ncdc depending on which proposal we go with for generated secrets in templates there may be an effect on this.   If secret creation does the encoding there is no problem, if templates have a base64 generator there will be a small impact

bparees commented:
@jwforres good point.  not sure what you'd want to do if we create a base64 generator...  would you try to recognize the generator type and decode the value?  or try to recognize the value itself as being base64 and decode it?

jhadvig commented:
@jwforres @spadgett @rhamilto I've updated the PR.
Based on the discussion we align on the fact we will hide the table by default with a link that will show it.
Also the sentence that describes the `Applied Parameter Values` section will list names of paramarets with generated values.
Screens:
![screenshot](https://cloud.githubusercontent.com/assets/1668218/15930706/9adc42b8-2e54-11e6-887d-c6bbf67d6437.png)

![screenshot-1](https://cloud.githubusercontent.com/assets/1668218/15930704/9751d43c-2e54-11e6-9bff-ebd4ac0b8d71.png)

In case of single generated parameter value
![screenshot-2](https://cloud.githubusercontent.com/assets/1668218/15930712/a18dbbb4-2e54-11e6-833e-1f023d2cd8cc.png)

jwforres commented on app/scripts/services/templates.js:
```diff
@@ -0,0 +1,53 @@
+    var params = [];
+    var paramsWithGeneratedValues = [];
+    return {
+      setParams: function(paramaters) {
```
misspelled parameters, couple places in this file

jwforres commented on app/scripts/services/templates.js:
```diff
@@ -0,0 +1,53 @@
+        _.each(paramaters, function(param) {
+          params.push({
+            name: param.name,
+            value: param.value
```
i think id rather see one pair of set/get methods and just have a 'wasGenerated' or 'generated' added to this map

jhadvig replied:
:+1: was thinking of it, but wanted to be as clear as possible.

jwforres commented on app/views/create/next-steps.html:
```diff
@@ -89,6 +89,18 @@ <h1 ng-if="!pendingTasks(tasks()).length && erroredTasks(tasks()).length">Comple
+            <div ng-if="processedParams.length">
+              <h2>Applied Parameter Values</h2>
+              <p>These parameters often include things like passwords. If you will need to reference these values later, copy them to a safe location.
+                <span ng-if="generatedParams.length > 1">Parameters <span ng-repeat="paramName in generatedParams">{{paramName}}{{$index+1 !== generatedParams.length ? "," : ""}} </span>were generated automatically.</span>
```
you can use $last to check whether to add the comma

```
<span ng-repeat="paramName in generatedParams">{{paramName}}<span ng-if="!$last">, </span></span>
```

jhadvig replied:
:ok_hand:

jwforres commented:
@bparees is it possible for us to know from the parameter info what type of generator it is? do not want to go down the road of detecting something is a base64 encoded value...

bparees commented:
@jwforres assuming we introduce a new type (like "base64_expression"), you can look at the generator type:
https://github.com/openshift/origin/blob/master/examples/quickstarts/nodejs-mongodb.json#L445

that field should still be available even after the template is processed (ie values generated).

jhadvig commented:
@jwforres I've addressed the your comments.
Once the introduce the new "base64_expression" like type, will update the logic. For now we should be ok.
PTAL :)

jwforres commented on app/scripts/services/templates.js:
```diff
@@ -0,0 +1,51 @@
+  }).service("ProcessedParametersService", function(){
+
+    var params = {
+      generated: [],
```
hmm these don't seem right, really its
params.all and params.generated

i was hoping it would have been a single array with generated being a flag on the individual params, but i see where that would complicate you changing from the one generated param to many generated params sentences in the view :-\

jwforres commented:
just fix the one comment and rebase, guessing the merge conflicts are on dist files

jhadvig commented:
@jwforres I've updated the PR, and put the generated dist files in a separate commit, cause they are huge for some unknown reason to me. Maybe its fine but would rather have your opinion on it.
Will squash afterwards :)

jwforres commented:
[merge]
